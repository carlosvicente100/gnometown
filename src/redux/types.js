// gnomes
export const fetchRequested = 'FETCH_REQUESTED';
export const fetchRequestedGnomes = 'FETCH_REQUESTED_GNOMES';
export const fetchPendingGnomes = 'FETCH_PENDING_GNOMES';
export const fetchSucceededGnomes = 'FETCH_SUCCEEDED_GNOMES';
export const fetchFailedGnomes = 'FETCH_FAILED_GNOMES';

//filters
export const setAgeFromTo = 'SET_BREWED_FROM_TO';
export const setAgeFrom = 'SET_BREWED_FROM';
export const setAgeTo = 'SET_BREWED_TO';
