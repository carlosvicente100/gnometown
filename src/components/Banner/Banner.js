import React from 'react'
import styled from 'styled-components'
import { useSelector } from 'react-redux'
import PropTypes from 'prop-types'
import { Link } from '@reach/router'

export const BannerContainer = styled.div`
  position: relative;
  background: white;
  width: 100%;
  height: 72px;
  display: grid;
  grid-template-columns: 15% 35% 35% 15%;
  grid-template-areas: '. m-town-name m-town-name m-filter';
  margin-bottom: 20px;

  label {
    text-align: center;
    grid-area: m-town-name;
    font-family: 'Helvetica';
    font-weight: 300;
    font-size: 34px;
    align-self: center;
    justify-self: center;
  }
`
export const Back = styled.div`
  font-size: 20px;
  margin: 20px;
`

const Banner = ({ isHome = false }) => {
  const { townName } = useSelector((state) => state.gnomesReducer)
  return (
    <BannerContainer>
      {!isHome && (
        <Back data-testid="banner-go-home">
          <Link to={`/`}>
            <i className="fas fa-home"></i>
          </Link>
        </Back>
      )}
      <label data-testid="banner-town-name">{`${townName} citizens`}</label>
    </BannerContainer>
  )
}

Banner.propTypes = {}

export default Banner
