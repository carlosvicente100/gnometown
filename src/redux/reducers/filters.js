import { fetchSetAgeFromTo, fetchSetJobList, fetchSetName, fetchResetFilters } from '../types'

export const initialState = {
  jobList: [],
  name: '',
  ageTo: {
    years: 0
  },
  ageFrom: {
    years: 0
  }
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchSetJobList:
      console.log(' +-+-+-+-+- action with jobs', action.data.jobs)
      return {
        ...state,
        jobList: action.data.jobs
      }
    case fetchSetName:
      console.log('filter by name', action.data.name)
      return {
        ...state,
        name: action.data.name
      }
    case fetchResetFilters:
      console.log('reset Filters')
      return {
        initialState
      }
    case fetchSetAgeFromTo:
      return {
        ...state,
        ageFrom: action.age.from,
        ageTo: action.age.to
      }
    default:
      return {
        ...state
      }
  }
}
