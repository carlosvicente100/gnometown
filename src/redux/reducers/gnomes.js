import { fetchPendingGnomes, fetchSucceededGnomes, fetchFailedGnomes } from '../types'

export const initialState = {
  pending: false,
  error: null,
  townName: '',
  gnomes: []
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchPendingGnomes:
      return {
        ...state,
        pending: true
      }
    case fetchSucceededGnomes:
      const townName = Object.keys(action.data)[0]
      return {
        ...state,
        pending: false,
        townName,
        gnomes: action.data[townName]
      }
    case fetchFailedGnomes:
      return {
        ...state,
        pending: false,
        error: true
      }
    default:
      return {
        ...state
      }
  }
}
