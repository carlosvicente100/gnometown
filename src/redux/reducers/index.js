import { combineReducers } from 'redux';
 import gnomesReducer from './gnomes';
import filtersReducer from './filters';

export default combineReducers({
  gnomesReducer,
  filtersReducer,
});
