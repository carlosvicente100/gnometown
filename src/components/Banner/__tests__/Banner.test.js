import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { render as rtlRender } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { initialState as initialStateGnomes } from '../../../redux/reducers/gnomes'
import reducer from '../../../redux/reducers'

import Banner from '../Banner'

beforeEach(() => {
  jest.clearAllMocks()
})

const initialReducerState = {
  gnomesReducer: initialStateGnomes
}

function render(ui, { initialState, store = createStore(reducer, initialState), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

const dummyProps = {
  isHome: true
}

describe('Banner Component Suite', () => {
  test('Snapshot check', () => {
    expect(render(<Banner />, { initialReducerState }).baseElement).toMatchSnapshot()
  })
})

describe('Banner Structure', () => {
  test('Should have a Label - townName by redux', () => {
    const dummyCityName = 'FunkyTown'
    const store = createStore(() => ({
      gnomesReducer: {
        townName: dummyCityName,
        pending: false,
        error: null,
        gnomes: []
      }
    }))
    const { getByTestId } = render(<Banner {...dummyProps} />, { store }) //banner-town-name
    const townName = getByTestId('banner-town-name')
    expect(townName.innerHTML).toBe(`${dummyCityName} citizens`)
  })

  test('Should have an icon to go Back - isHome prop', () => {
    const dummyCityName = 'FunkyTown'
    const store = createStore(() => ({
      gnomesReducer: {
        townName: dummyCityName,
        pending: false,
        error: null,
        gnomes: []
      }
    }))
    const { getByTestId } = render(<Banner isHome={false} />, { store }) //banner-town-name
    const backDiv = getByTestId('banner-go-home')
    const HomeLink = backDiv.children[0]
    expect(HomeLink.getAttribute('href')).toBe('/')
    expect(HomeLink.children[0].getAttribute('class')).toBe('fas fa-home')
  })
})
