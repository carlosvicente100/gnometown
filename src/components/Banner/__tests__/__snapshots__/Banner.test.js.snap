// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner Component Suite Snapshot check 1`] = `
.c0 {
  position: relative;
  background: white;
  width: 100%;
  height: 72px;
  display: grid;
  grid-template-columns: 15% 35% 35% 15%;
  grid-template-areas: '. m-town-name m-town-name m-filter';
  margin-bottom: 20px;
}

.c0 label {
  text-align: center;
  grid-area: m-town-name;
  font-family: 'Helvetica';
  font-weight: 300;
  font-size: 34px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
}

.c1 {
  font-size: 20px;
  margin: 20px;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="banner-go-home"
      >
        <a
          aria-current="page"
          href="/"
        >
          <i
            class="fas fa-home"
          />
        </a>
      </div>
      <label
        data-testid="banner-town-name"
      >
         citizens
      </label>
    </div>
  </div>
</body>
`;
